### ISSUE TITLE
v 3.2.7 conflicts with Umatrix Javascript Blocking

### ISSUE THREAD

**Issue by Copyrites** [Cimbali/CleanLinks#67]
I use uMatrix to block all javascript in google.com search results.

My browser (firefox 67.0.1 64-bit ubuntu) upgraded to CleanLinks 3.2.7 and now google search results are being improperly cleaned.    Reverting to 3.2.6 fixes the problem.    Setting uMatrix to allow javascript on google.com also fixes the problem (but obviously creates all kinds of other problems).

I verified this behavior on a brand new profile, no other extensions installed.

It looks like the leading URL is getting stripped, but not the extra google-specific parameters.  For example:

https://www.google.com/url?q=https://www.urbandictionary.com/define.php%3Fterm%3Dnewbie&sa=U&ved=2ahUKEwizjufMq-LiAhXoqlkKHZvXD0UQFjADegQIBxAB&usg=AOvVaw0Fh9Evi05dusQaHVvgTzH6

Gets turned into:

https://www.urbandictionary.com/define.php%3Fterm%3Dnewbie&sa=U&ved=2ahUKEwizjufMq-LiAhXoqlkKHZvXD0UQFjADegQIBxAB&usg=AOvVaw0Fh9Evi05dusQaHVvgTzH6

**Comment by Cimbali** [comment]
This behaviour comes from the fix for #64 which alllows to detect raw (non-encoded) links in redirects. There is an inconsistency in the link you show, where the `:` and `/`are not encoded for some reason, whereas all the next tokens are encoded such as `?` as `%3F`. It is then impossible to detect the start and end of the URL.

However, from what I can see in google results for the term newbie (with JS enabled), the original URL is:

> https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=6&ved=2ahUKEwjslryHu-LiAhXQyYUKHfgaDEsQFjAFegQIBRAB&url=https%3A%2F%2Fwww.urbandictionary.com%2Fdefine.php%3Fterm%3Dnewbie&usg=AOvVaw0C8_1j9LKyaip4KavmdSGf

So as you can see the `:` and `/` are correctly encoded as `%3A` and `%2F` and, unsurprisingly, the cleaned URL is correct:

> https://www.urbandictionary.com/define.php?term=newbie

Strangely when using uMatrix to remove javascript I see a clean link, and not a dirty one, even with all other addons disabled. However in the firefox session I use to debug, the links appear dirty and inconsistently encoded like yours. I even purged both caches and still get different results across both instances of the same browser with the same add-ons enabled… So I’m not sure where it comes from yet and I’ll look into it.

Finally, and as a temporary work-around (alternately to reverting to v3.2.6), there is an add-on called “Google Search URL Fixup” which is the second hit for “uMatrix” on addons.mozilla.org, which replaces the links on the google page with the cleaned ones.

**Comment by Copyrites** [comment]
Maybe related: removal of the ?ref_= parameter on imdb.com stopped working with 3.2.7, reverting to 3.2.6 fixes it.  This behavior happens even after disabling umatrix completely.

Example URL (paste into URL bar): https://www.imdb.com/title/tt7575778/?ref_=fn_al_tt_1

3.2.7 does not modify the URL
3.2.6 changes to https://www.imdb.com/title/tt7575778/

Verified in the same fresh profile as above (with umatrix disabled but not removed)

**Comment by Cimbali** [comment]
See #68

**Comment by Cimbali** [comment]
So interestingly there is a `noscript` tag that redirects you from the google page with correct links to the one with inconsistent links.

```
<noscript>
	<meta content="0;url=/search?q=newbie&amp;gbv=1&amp;sei=IXICXaTuGanJgwf57aHICg" http-equiv="refresh">
	<style>table,div,span,p{display:none}</style>
	<div style="display:block">Por favor, haz clic <a href="/search?q=newbie&amp;gbv=1&amp;sei=IXICXaTuGanJgwf57aHICg">aquí</a> si no has sido redireccionado en unos segundos.</div>
</noscript>
```

It also adds some parameters (`sei`, `gbv`) to the next page, which is then the no-javascript specialized page, which:
- has the incorrectly-encoded links,
- has different styling,
- sets some cookies so that all your next interactions with google happen on this type of no-javascript specialized page even without the supplementary parameters.

  (Blocking cookies causes this redirection with parameters to happen every time rather than the no-javascript page being loaded immediately.)

There is a lot if literature on uMatrix and google results interactions due to noscript tags, such as gorhill/uMatrix#232 and gorhill/uMatrix#319. From what I gather, it’s not possible to just block `noscript` tags as well as javascript, because this means empty pages on image results. So you have to live with some cookie-based tracking mechanism and CleanLinks will have to live with handling those links correctly.

I suppose mixed-encoding (as no-javascript google does) URLs is (slightly?) more common than the use case in #64, so when we’ll have per-domain rules I’ll roll back this part of release v3.2.7 and make it domain-specific. I’m really not sure I can find a heuristic because for any link as the one below, it is impossible to determine if the `param` argument belongs to the base or nested URL:

    http://somedomain.com?arg=http://destinat.ion/&param=val

In the mean time, I can only recommend using the additional add-on [Google Search URL Fixup](https://addons.mozilla.org/en-GB/firefox/search/?q=Google%20Search%20URL%20Fixup&platform=Linux&appver=66.0.5) or rolling back to v3.2.6.

**Comment by Cimbali** [comment]
The new cleaning mechanism that will be released with version v4.0 seems to fix this use case. You can test it at: https://cimbali.github.io/CleanLinks/

From your example link, I get `https://www.urbandictionary.com/define.php?term=newbie`
